pkg/sharky/store.go

```diff
@@ -0,0 +1,190 @@
+	file, err := basedir.Open(fmt.Sprintf("shard_%03d", index))
+	if err != nil {
+		return nil, err
+	}
+	ffile, err := basedir.Open(fmt.Sprintf("free_%03d", index))
+	if err != nil {
+		return nil, err
+	}
+	sl := newSlots(ffile.(sharkyFile), s.wg)
+	err = sl.load()
+	if err != nil {
+		return nil, err
+	}
+	sh := &shard{
+		reads:       make(chan read),
+		errc:        make(chan error),
+		writes:      s.writes,
+		index:       index,
+		maxDataSize: maxDataSize,
+		file:        file.(sharkyFile),
+		slots:       sl,
+		quit:        s.quit,
+	}
+	terminated := make(chan struct{})
+	sh.slots.wg.Add(1)
+	go func() {
+		sh.process()
+		close(terminated)
+	}()
+	go sl.process(terminated)
+	return sh, nil
+}
+
+// Read reads the content of the blob found at location into the byte buffer given
+// The location is assumed to be obtained by an earlier Write call storing the blob
+func (s *Store) Read(ctx context.Context, loc Location, buf []byte) (err error) {
+	sh := s.shards[loc.Shard]
+	select {
+	case sh.reads <- read{buf[:loc.Length], loc.Slot}:
+		s.metrics.TotalReadCalls.Inc()
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+
+	select {
+	case err = <-sh.errc:
+		if err != nil {
+			s.metrics.TotalReadCallsErr.Inc()
+		}
+		return err
+	case <-s.quit:
+		return ErrQuitting
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+// Write stores a new blob and returns its location to be used as a reference
+// It can be given to a Read call to return the stored blob.
+func (s *Store) Write(ctx context.Context, data []byte) (loc Location, err error) {
+	if len(data) > s.maxDataSize {
+		return loc, ErrTooLong
+	}
+	s.wg.Add(1)
+	defer s.wg.Done()
+
+	c := make(chan entry)
+
+	select {
+	case s.writes <- write{data, c}:
+		s.metrics.TotalWriteCalls.Inc()
+	case <-s.quit:
+		return loc, ErrQuitting
+	case <-ctx.Done():
+		return loc, ctx.Err()
+	}
+
+	select {
+	case e := <-c:
+		if e.err == nil {
+			shard := strconv.Itoa(int(e.loc.Shard))
+			s.metrics.CurrentShardSize.WithLabelValues(shard).Add(1)
+		} else {
+			s.metrics.TotalWriteCallsErr.Inc()
+		}
+		return e.loc, e.err
+	case <-s.quit:
+		return loc, ErrQuitting
+	case <-ctx.Done():
+		return loc, ctx.Err()
+	}
+}
+
+// Release gives back the slot to the shard
+// From here on the slot can be reused and overwritten
+// Release is meant to be called when an entry in the upstream db is removed
+// Note that releasing is not safe for obfuscating earlier content, since
+// even after reuse, the slot may be used by a very short blob and leaves the
+// rest of the old blob bytes untouched
+func (s *Store) Release(ctx context.Context, loc Location) error {
+	sh := s.shards[loc.Shard]
+	// we add the current routine and will be Done in slots.process
+	sh.slots.wg.Add(1)
+	err := sh.release(ctx, loc.Slot)
+	s.metrics.TotalReleaseCalls.Inc()
+	if err == nil {
+		shard := strconv.Itoa(int(sh.index))
+		s.metrics.CurrentShardSize.WithLabelValues(shard).Sub(1)
```

Comment:
so this should hold the size of shard in actual stored chunks... but it might be also useful to somehow keep track of the actual size of the shard on the disk (i'm not sure if this can be done in some way which is not too ugly since you might need to leak stuff from the slots component). this would allow us to know for example if there's any leakage of the shard (i.e. shard keeps growing while there are free slots), etc